src/metpy/calc/boundarylayer.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/python
+# -*-coding:utf-8 -*-
+"""
+Contains a collection of boundary layer height estimations.
+
+
+References:
+-----------
+
+[Col14]: Collaud Coen, M., Praz, C., Haefele, A., Ruffieux, D., Kaufmann, P., and Calpini, B. (2014): 
+    Determination and climatology of the planetary boundary layer height above the Swiss plateau by in situ and remote sensing measurements as well as by the COSMO-2 model
+    Atmos. Chem. Phys., 14, 13205–13221.
+
+[HL06]: Hennemuth, B., & Lammert, A. (2006):
+    Determination of the atmospheric boundary layer height from radiosonde and lidar backscatter.
+    Boundary-Layer Meteorology, 120(1), 181-200.
+
+[Guo16]: Guo, J., Miao, Y., Zhang, Y., Liu, H., Li, Z., Zhang, W., ... & Zhai, P. (2016):
+    The climatology of planetary boundary layer height in China derived from radiosonde and reanalysis data.
+    Atmos. Chem. Phys, 16(20), 13309-13319.
+
+[Sei00]: Seidel, D. J., Ao, C. O., & Li, K. (2010):
+    Estimating climatological planetary boundary layer heights from radiosonde observations: Comparison of methods and uncertainty analysis.
+    Journal of Geophysical Research: Atmospheres, 115(D16).
+    
+[VH96]: Vogelezang, D. H. P., & Holtslag, A. A. M. (1996):
+    Evaluation and model impacts of alternative boundary-layer height formulations.
+    Boundary-Layer Meteorology, 81(3-4), 245-269.
+"""
+
+import numpy as np
+from copy import deepcopy
+
+import metpy.calc as mpcalc
+import metpy.constants as mpconsts
+from metpy.units import units
+
+
+def smooth(val, span):
```

Comment:
Thanks for the references! I knew some equivalent functions were already existing but they are not quite exactly the same (Xarray works on xarray.Dataset, Scipy has a slightly different strategy at the edges) and, given that the function is simple enough, it was less work to write it than to look for the existing one.

Bottleneck's function seems to do exactly what I want but it is not listed in the Metpy's dependencies. Do you think it's worth adding it so I can use their moving-mean function?

Reply:
You'd have to ask one of the maintainers, but, in the meantime, would this be faster?
```python
cumulative_sums = np.nancumsum(val)
rolling_sums = cumulative_sums[span:] - cumulative_sums[:-span]
valid_index = np.isfinite(val)
cumulative_count = np.cumsum(valid_index)
rolling_count = cumulative_count[span:] - cumulative_count[:-span]
rolling_means = rolling_sums / rolling_count
```
You'd need to pre-allocate `rolling_means` and handle the edges still but it should work.  (Alternately, use [`np.lib.stride_tricks.sliding_window_view`](https://numpy.org/devdocs/reference/generated/numpy.lib.stride_tricks.sliding_window_view.html) with `np.nanmean`, but the note about it being slow is warranted)

Alternately, use SciPy for the bulk of the calculation, then re-do the edges the way you want.

It would probably be a good idea to check whether this takes enough time that it's worth optimizing before going too far, though (as you may have noticed, I am not good at that).

Reply:
With the testing data I have it's almost instantaneous and, as I moved other topics, I don't really have something bigger to quickly try it on. I suggest we leave it that way for now and other users might open another issue if when need to speed it up. Would that be alright?

Reply:
You picked the same name as [elsewhere in MetPy](https://github.com/Unidata/MetPy/blob/6924706ba8e695d7ced1d7faa279593f6ae841ea/src/metpy/calc/basic.py#L948-L978), though the edge handling is again different (they do not smooth close to the edge).

The bigger test data would likely be someone trying to find boundary layer height from model data somewhere, and it looks like your code is designed for a profile at a time, not arrays of profiles (either the N x Z that description implies or the Z x Y x X conventional in model output).  It might be a simple matter to add an `axis` keyword argument to most functions and borrow [the logic from the derivative functions to create the indexers](https://github.com/Unidata/MetPy/blob/6924706ba8e695d7ced1d7faa279593f6ae841ea/src/metpy/calc/tools.py#L1240) (and possibly the [vertical axis auto-detection for DataArrays](https://github.com/Unidata/MetPy/blob/6924706ba8e695d7ced1d7faa279593f6ae841ea/src/metpy/xarray.py#L575)), but that should probably be a follow-up PR in case it isn't.